src/experimental/tools/VideoThumbnailLoader/thumbnail_loader.ts

```diff
@@ -0,0 +1,306 @@
+import {
+  IContentInfos,
+  ILoaders,
+} from "./types";
+import VideoThumbnailLoaderError from "./video_thumbnail_loader_error";
+
+const PPromise = typeof Promise === "function" ? Promise :
+                                                 pinkie;
+
+const MIN_NEEDED_DATA_AFTER_TIME = 2;
+
+interface ITimeSettingTask { contentInfos: IContentInfos;
+                             time: number;
+                             stop: () => void;
+                             promise: Promise<number>; }
+
+/**
+ * This tool, as a supplement to the RxPlayer, intent to help creating thumbnails
+ * from a video source.
+ *
+ * The tools will extract a "thumbnail track" either from a video track (whose light
+ * chunks are adapted from such use case) or direclty from the media content.
+ */
+export default class VideoThumbnailLoader {
+  private readonly _videoElement: HTMLVideoElement;
+
+  private _player: Player;
+  private _currentTask?: ITimeSettingTask;
+  private _nextTaskSegments?: ISegment[];
+  private _loaders: ILoaders = {};
+  private _requests: Map<string, ICancellableRequest>;
+  constructor(videoElement: HTMLVideoElement,
+              player: Player) {
+    this._videoElement = videoElement;
+    this._player = player;
+    this._requests = new Map<string, ICancellableRequest>();
+  }
+
+  /**
+   * Add imported loader to thumbnail loader loader object.
+   * It allows to use it when setting time.
+   * @param {function} loaderFunc
+   */
+  addLoader(loaderFunc: (features: ILoaders) => void): void {
+    loaderFunc(this._loaders);
+  }
+
+  /**
+   * Set time of thumbnail video media element :
+   * - Remove buffer when too much buffered data
+   * - Search for thumbnail track element to display
+   * - Load data
+   * - Append data
+   * Resolves when time is set.
+   * @param {number} time
+   * @returns {Promise}
+   */
+  setTime(time: number): Promise<number> {
+    if (this._currentTask !== undefined) {
+      if (time === this._currentTask.time) {
+        // The current task is already handling the loading for the wanted time
+        // (same thumbnail).
+        return this._currentTask.promise;
+      }
+    }
+
+    const manifest = this._player.getManifest();
+    if (manifest === null) {
+      return PPromise.reject(
+        new VideoThumbnailLoaderError("NO_MANIFEST",
+                                      "No manifest available."));
+    }
+    const contentInfos = getContentInfos(time, manifest);
+    if (contentInfos === null) {
+      return PPromise.reject(
+        new VideoThumbnailLoaderError("NO_TRACK",
+                                      "Couldn't find track for this time."));
+    }
+
+    const segments = contentInfos
+      .representation.index.getSegments(time, MIN_NEEDED_DATA_AFTER_TIME);
+
+    if (segments.length === 0) {
+      return PPromise.reject(
+        new VideoThumbnailLoaderError("NO_THUMBNAIL",
+                                      "Couldn't find thumbnail."));
```

Comment:
maybe adding "for the given time" would be more indicative. Here we don't know if that's in general or what